packages/std/src/errors/std_error.rs

```diff
@@ -0,0 +1,313 @@
+                },
+            ) => source == source2 && msg == msg2 && backtraces_eq(backtrace, backtrace2),
+            (
+                StdError::Unauthorized { backtrace },
+                StdError::Unauthorized {
+                    backtrace: backtrace2,
+                },
+            ) => backtraces_eq(backtrace, backtrace2),
+            (
+                StdError::Underflow {
+                    minuend,
+                    subtrahend,
+                    backtrace,
+                },
+                StdError::Underflow {
+                    minuend: minued2,
+                    subtrahend: subtrahend2,
+                    backtrace: backtrace2,
+                },
+            ) => {
+                minuend == minued2
+                    && subtrahend == subtrahend2
+                    && backtraces_eq(backtrace, backtrace2)
+            }
+            _ => false,
+        }
+    }
+}
+
+/// The return type for init, handle and query. Since the error type cannot be serialized to JSON,
+/// this is only available within the contract and its unit tests.
+///
+/// The prefix "Std" means "the standard result within the standard library". This is not the only
+/// result/error type in cosmwasm-std.
+pub type StdResult<T> = core::result::Result<T, StdError>;
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::serde::{from_slice, to_vec};
+
+    #[test]
+    fn can_serialize() {
+        let error = InvalidBase64 {
+            msg: "invalid length".to_string(),
+        }
+        .build();
+        assert_eq!(
+            to_vec(&error).unwrap(),
+            br#"{"invalid_base64":{"msg":"invalid length"}}"#.to_vec()
+        );
+    }
+
+    #[test]
+    fn can_deserialize() {
+        let error: StdError =
+            from_slice(br#"{"invalid_base64":{"msg":"invalid length"}}"#).unwrap();
+        match error {
+            StdError::InvalidBase64 { msg, backtrace } => {
+                assert_eq!(msg, "invalid length");
+                assert!(backtrace.is_none());
+            }
+            _ => panic!("invalid type"),
+        };
+    }
+
+    /// The deseralizer in from_slice can perform zero-copy deserializations (https://serde.rs/lifetimes.html).
+    /// So it is possible to have `&'static str` fields as long as all source data is always static.
+    /// This is an unrealistic assumption for our use case. This test case ensures we can deseralize
+    /// errors from limited liefetime sources.
+    #[test]
+    fn can_deserialize_from_non_static_source() {
+        let source = (br#"{"not_found":{"kind":"bugs"}}"#).to_vec();
+        let error: StdError = from_slice(&source).unwrap();
+        match error {
+            StdError::NotFound { kind, backtrace } => {
+                assert_eq!(kind, "bugs");
+                assert!(backtrace.is_none());
+            }
+            _ => panic!("invalid type"),
+        };
+    }
+
+    #[test]
+    fn eq_works() {
```

Comment:
huh?

This tests `ne_works`

Reply:
Also, there is a cleaner way to avoid backtraces than serializing and deserilizing - simply set:

```rust
let error1 = StdError::InvalidBase64 {
  msg: "invalid length".to_string(),
  backtrace: None,
};
```

that makes it clearer what is being tested

Reply:
Well spotted. Fixed and simplified in https://github.com/CosmWasm/cosmwasm/pull/305/commits/51fbdb64ce0cfb24b9d3f1a99821773153fe13b1